python/dgl/distributed/kvstore.py

```diff
@@ -0,0 +1,1107 @@
+        """
+        if self._client_id == 0:
+            request = RegisterPushHandlerRequest(func)
+            # send request
+            for server_id in range(self._server_count):
+                rpc.send_request(server_id, request)
+            # wait confirmation
+            for _ in range(self._server_count):
+                response = rpc.recv_response()
+
+    def register_pull_handler(self, func):
+        """Register UDF pull function on server.
+
+        client_0 will send this request to all server.
+
+        Parameters
+        ----------
+        func : UDF pull function
+        """
+        if self._client_id == 0:
+            request = RegisterPullHandlerRequest(func)
+            # send request
+            for server_id in range(self._server_count):
+                rpc.send_request(server_id, request)
+            # wait confirmation
+            for _ in range(self._server_namebook):
+                response = rpc.recv_response()
+
+    def get_shared_data(self, partition_book):
+        """Mapping shared-tensor from server to client.
+
+        Parameters
+        ----------
+        partition_book : GraphPartitionBook or RangePartitionBook
+            Store the partition information
+        """
+        # Get meta data
+        request = GetSharedDataRequest(self._client_id)
+        rpc.send_request(self._main_server_id, request)
+        response = rpc.recv_response()
+        for name, meta in res.meta.items():
+            shape, dtype, plolicy_str = meta
+            shared_data = empty_shared_mem(name, False, shape, dtype)
+            dlpack = shared_data.to_dlpack()
+            self._data_store[name] = F.zerocopy_from_dlpack(dlpack)
+            self._part_policy[name] = PartitionPolicy(policy_str, self._part_id, partition_book)
+        # Get full data shape
+        for name, data in self._data_store.items():
+            data_shape = list(F.shape(data))
+            data_shape[0] = 0
+            request = GetFullShapeRequest(name)
+            # send request
+            for machine_id in range(self._machine_count):
+                server_id = machine_id * self._group_count
+                rpc.send_request(server_id, request)
+            # recv response
+            for _ in range(self._machine_count):
+                response = rpc.recv_response()
+                data_shape[0] += response.shape[0]
+            self._full_data_shape[name] = tuple(data_shape)
+
+    def barrier(self):
+        """Barrier for all client nodes
+
+        This API will be blocked untill all the clients call this API.
+        """
+        request = BarrierRequest(self._client_id)
+        # send request
+        for server_id in range(self._server_count):
+            rpc.send_request(server_id, request)
+        # recv response
+        for _ in range(self._server_count):
+            response = rpc.recv_response()
+
+    def init_data(self, name, shape, dtype, policy_str, partition_book, init_func):
```

Comment:
i think you just need to pass partition policy to kvstore.

Reply:
I don't think so. We also need to set the partition policy on the local client. KVStore can't mapping partition_book from server to client.

Reply:
you only need to pass the partition policy for the local partition. i think passing a partition policy should be sufficient.

Reply:
But we need the parition_book to calculate the shape of other partitions.